Add QLinearSigmoid and QLinearLekyRelu op

gyulaz-htec opened the pull request:
[QLinearSigmoid](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.QLinearSigmoid) operator is required for [Faster-RCNN-int8](https://github.com/onnx/models/blob/main/vision/object_detection_segmentation/faster-rcnn/model/FasterRCNN-12-int8.onnx) and [Mask-RCNN-int8](https://github.com/onnx/models/blob/main/vision/object_detection_segmentation/mask-rcnn/model/MaskRCNN-12-int8.onnx) onnx zoo models.

Fixes: https://github.com/migraphx-benchmark/AMDMIGraphX/issues/151 and https://github.com/migraphx-benchmark/AMDMIGraphX/issues/156

pfultz2 commented on src/onnx/parse_qlinearsigmoid.cpp:
```diff
@@ -0,0 +1,129 @@
+
+struct parse_qlinearsigmod : op_parser<parse_qlinearsigmod>
+{
+    std::vector<op_desc> operators() const { return {{"QLinearSigmoid"}}; }
```
You can do `{{"QLinearSigmoid", "sigmoid"}}`

gyulaz-htec commented on src/onnx/parse_qlinearsigmoid.cpp:
```diff
@@ -0,0 +1,129 @@
+
+*/
+
+struct parse_qlinearsigmod : op_parser<parse_qlinearsigmod>
```
There is currently only one other unary qlineary operator in migraphx [QLinearglobalAveragePool](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.QLinearGlobalAveragePool), but that requires more detailed input checking (it has no optional inputs) also calling the underlying migraphx operator for that is more complex than a simple  `add_common_op` call.
On the other hand I'm planning to add `QLinearLeakyRelu` and `QLinearAveragePool` ops which are unary operators with similar requirements. I will try to update this to `parse_qlinearunary`.

pfultz2 replied:
`QLinearAveragePool` should be handled by a different parser, but the `QLinearLeakyRelu` can reuse the same parse.

gyulaz-htec commented:
@pfultz2 I've updated the PR added comon parsing of qlinearunary ops and moved them in  the same file. Also Added QLinearLeakyRelu to the batch, because it is also required for ONNX zoo models and is a unary op.

pfultz2 commented on src/onnx/parse_qlinearunary.cpp:
```diff
@@ -0,0 +1,299 @@
+}
+
+template <class F>
+instruction_ref parse_qlinearunary(const onnx_parser::node_info& info,
```
Move this function to a seperate header, similiar to the other helper functions.

pfultz2 replied:
Also this function should be named `add_qlinearunary` not with `parse_` which is generally reserved for the parser classes.

gyulaz-htec replied:
Removed this function because the change you mentioned below makes this solution obsolete.

pfultz2 commented on src/onnx/parse_qlinearunary.cpp:
```diff
@@ -0,0 +1,299 @@
+
+*/
+
+struct parse_qlinearsigmod : op_parser<parse_qlinearsigmod>
```
This should be name `parse_qlinearunary`.

gyulaz-htec replied:
Renamed it to parse_qlinearunary

pfultz2 commented on src/onnx/parse_qlinearunary.cpp:
```diff
@@ -0,0 +1,299 @@
+
+struct parse_qlinearsigmod : op_parser<parse_qlinearsigmod>
+{
+    std::vector<op_desc> operators() const { return {{"QLinearSigmoid", "sigmoid"}}; }
```
This should be `{"QLinearSigmoid", "sigmoid"}, {"QLinearLeakyRelu", "leaky_relu"}}`.

pfultz2 commented on src/onnx/parse_qlinearunary.cpp:
```diff
@@ -0,0 +1,299 @@
+    {
+        check_qlinearunary_inputs(opd, args);
+        return parse_qlinearunary(info, args, [&](instruction_ref dquant_x) {
+            return info.add_instruction(migraphx::make_op(opd.op_name), dquant_x);
```
This should do:

```cpp
auto op = parser.load(opd.op_name, info);
return info.add_instruction(op, dquant_x);
```

So it works for leaky_relu as well.
```

pfultz2 commented on src/onnx/parse_qlinearunary.cpp:
```diff
@@ -0,0 +1,299 @@
+
+*/
+
+struct parse_qlinearglobalaveragepool : op_parser<parse_qlinearglobalaveragepool>
```
This should be name `qlinearpool` and it should be moved to a separate cpp file.

gyulaz-htec replied:
This was previously in the deleted `parse_qlinearglavgpool.cpp`, but if this should be in a separate file I will restore `parse_qlinearglavgpool.cpp`. As a follow up I can change it to `parse_qlinearpooling` when adding QLinearAveragePool.

gyulaz-htec commented:
@pfultz2 I've made the requested changes, as I've mentioned in one of the comments I've reset the `parse_qlinearglavgpool.cpp` and moving it under parse_qlinearpooling will be done in a seperate PR.

gyulaz-htec commented:
@causten the Performance results seem to be okay now.
